spyder/plugins/ipythonconsole/widgets/shell.py

```diff
@@ -369,30 +375,79 @@ def setup_spyder_kernel(self):
         self.kernel_handler.poll_fault_text()
 
-        # Show possible errors when setting Matplotlib backend
-        self.call_kernel().show_mpl_backend_errors()
+        self.send_spyder_kernel_configuration()
 
-        # Check if the dependecies for special consoles are available.
-        self.call_kernel(
-            callback=self.ipyclient._show_special_console_error
-            ).is_special_kernel_valid()
+        run_lines = self.get_conf('startup/run_lines')
+        if run_lines:
+            self.execute(run_lines, hidden=True)
 
-        self.send_spyder_kernel_configuration()
+        if self.get_conf('startup/use_run_file'):
+            run_file = self.get_conf('startup/run_file')
+            if run_file:
+                self.execute(f"exec(open({run_file}))", hidden=True)
 
     def send_spyder_kernel_configuration(self):
         """Send kernel configuration to spyder kernel."""
+        self.is_kernel_configured = False
+
+        # Set matplotlib backend
+        self.send_mpl_backend()
+
+        # set special kernel
+        self.set_special_kernel()
+
         # Set current cwd
         self.set_cwd()
 
         # To apply style
         self.set_color_scheme(self.syntax_style, reset=False)
 
         # Enable faulthandler
-        self.kernel_handler.enable_faulthandler()
+        self.set_kernel_configuration("faulthandler", True)
 
         # Give a chance to plugins to configure the kernel
         self.sig_config_spyder_kernel.emit()
 
+        if self.is_external_kernel:
+            # Enable wurlitzer
+            # Not necessary if started by spyder
+            # Does not work if the external kernel is on windows
+            self.set_kernel_configuration("wurlitzer", True)
+
+        if self.get_conf('autoreload'):
+            # Enable autoreload_magic
+            self.set_kernel_configuration("autoreload_magic", True)
+
+        self.call_kernel(
+            interrupt=self.is_debugging(),
+            callback=self.kernel_configure_callback
+        ).set_configuration(self._kernel_configuration)
+
+        self.is_kernel_configured = True
+
+    def set_kernel_configuration(self, key, value):
+        """Set kernel configuration."""
+        if self.is_kernel_configured:
+            if (
+                key not in self._kernel_configuration
+                or self._kernel_configuration[key] != value
+            ):
+                # Do not send twice
+                self.call_kernel(
+                    interrupt=self.is_debugging(),
+                    callback=self.kernel_configure_callback
+                ).set_configuration({key: value})
+
+        self._kernel_configuration[key] = value
+
+    def kernel_configure_callback(self, dic):
+        """Kernel configuration callback"""
+        for key, value in dic.items():
+            if key == "faulthandler":
+                self.kernel_handler.faulthandler_setup(value)
+            elif key == "special_kernel":
```

Comment:
Because this is a reply the key if the same as the one sent by the frontend. So you ask the kernel to set some special kernel, and if it fails, it will reply with the appropriate error